src/lib/pipe_lib/broadcast_pipe.ml

```diff
@@ -0,0 +1,134 @@
+open Core_kernel
+open Async_kernel
+
+type 'a t =
+  { mvar: 'a option Mvar.Read_only.t
+  ; mutable cache: 'a option
+  ; mvar_pipe_handle: 'a option Pipe.Reader.t
+  ; mutable id: int
+  ; pipes:
+      ( 'a option Strict_pipe.Reader.t
+      * ( 'a option
+        , Strict_pipe.drop_head Strict_pipe.buffered
+        , unit )
+        Strict_pipe.Writer.t )
+      Int.Table.t }
+
+let create mvar =
+  let mvar_pipe_handle = Mvar.pipe_when_ready mvar in
+  let t =
+    { mvar
+    ; cache= Mvar.peek mvar |> Option.join
+    ; mvar_pipe_handle
+    ; id= 0
+    ; pipes= Int.Table.create () }
+  in
+  don't_wait_for
+    (Pipe.iter_without_pushback t.mvar_pipe_handle ~f:(fun a ->
+         t.cache <- a ;
+         Int.Table.iteri t.pipes ~f:(fun ~key ~data:(_, w) ->
+             Core.printf !"Pipe%d %{sexp: int option}\n%!" key (Obj.magic a) ;
+             Strict_pipe.Writer.write w a ) )) ;
+  t
+
+let peek {cache; _} = cache
+
+let close t =
+  Core.printf !"Close all\n%!" ;
```

Comment:
Remove the printf. Maybe you should have a log instead

Reply:
Good catch, I left debug code in here